### DISCUSSION TITLE
Flat Index Use GPU search

### DISCUSSION THREAD

**Discussion by 1556900941lizerui** [milvus-io/milvus#37042] [Q&A and General discussion]
Does the FLAT index support GPU search? If so, does it support the COSINE calculation method?And How to use it by Python

**Comment by yhmo** [comment]
Supported GPU indexes are listed here: https://milvus.io/docs/gpu_index.md#GPU-Index

No GPU_FLAT.
There is a search parameter "nprobe" for GPU_IVF_FLAT, you can set "nprobe" to be equal to "nlist" to achieve a full scan like FLAT.

```
collection.create_index(field_name = "vector",
          index_params = {"index_type": "GPU_IVF_FLAT", "metric_type": "COSINE", "params": {"nlist": 64}}
)
```

```
search_params = {
        "metric_type": "COSINE",
        "params": {"nprobe": 64}
    }
res = collection.search(
    data=vectors,
    anns_field="vector",
    limit=limit,
    param=search_params,
)
```

**Comment by 1556900941lizerui** [comment]
OK,Thank you.



------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2024年10月22日(星期二) 下午2:46
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [milvus-io/milvus] Flat Index Use GPU search (Discussion #37042)














 
Supported GPU indexes are listed here: https://milvus.io/docs/gpu_index.md#GPU-Index
 
No GPU_FLAT.
 There is a search parameter "nprobe" for GPU_IVF_FLAT, you can set "nprobe" to be equal to "nlist" to achieve a full scan like FLAT.
 
collection.create_index(field_name = "my_vector", index_params = {"index_type": "GPU_IVF_FLAT", "metric_type": "COSINE", "params": {"nlist": 64}})
 search_params = {         "metric_type": "COSINE",         "params": {"nprobe": 64}     } res = collection.search(     data=vectors,     anns_field="vector",     limit=limit,     param=search_params, )

**Comment by 1556900941lizerui** [comment]
Currently, the number of items in my collection is only tens of thousands at most. What are the appropriate values ​​for nlist and nprobe? When nprobe and nlist of GPU_IVF_FLAT are the same, it is completely equivalent to FLAT.



------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "milvus-io/milvus"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年10月22日(星期二) 下午2:46
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [milvus-io/milvus] Flat Index Use GPU search (Discussion #37042)





 
Supported GPU indexes are listed here: https://milvus.io/docs/gpu_index.md#GPU-Index
 
No GPU_FLAT.
 There is a search parameter "nprobe" for GPU_IVF_FLAT, you can set "nprobe" to be equal to "nlist" to achieve a full scan like FLAT.
 
collection.create_index(field_name = "my_vector", index_params = {"index_type": "GPU_IVF_FLAT", "metric_type": "COSINE", "params": {"nlist": 64}})
 search_params = {         "metric_type": "COSINE",         "params": {"nprobe": 64}     } res = collection.search(     data=vectors,     anns_field="vector",     limit=limit,     param=search_params, )

**Reply by yhmo** [reply]
https://milvus.io/docs/performance_faq.md#How-to-set-nlist-and-nprobe-for-IVF-indexes
 `As a rule of thumb, the recommended value of nlist is 4 × sqrt(n), where n is the total number of entities in a segment.`

The collection has only tens of thousands items, I think there is only one segment in the collection. Let's say the collection has 50000 items, the proper value of nlist is 4 * sqrt(50000) = 894. Usually we set the value as power of 2. So, we can set it to be 1024.

If the collection has millions of items, then we need to estimate row count per segment by vector's dimension.

**Comment by 1556900941lizerui** [comment]
Thank you. I still need to ask some questions. Could you please give me your answer?



1. I currently have a data of about 10,000 entities, and I think the FLAT index is completely sufficient in terms of performance. When I use the Attu visualization tool to perform vector retrieval, I find that it only takes tens of milliseconds (the red box in the picture shows the time, I am not sure whether this represents the retrieval time), but when I use the clinet.search() of the Python SDK When running the function, it is found that the time is between 100ms and 500ms, and the time will keep fluctuating. I don't know why, so I'm thinking about using a GPU Speed ​​up retrieval.



2. Will CPU performance affect the time of vector retrieval for data of my magnitude?




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "milvus-io/milvus"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年10月23日(星期三) 下午3:08
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [milvus-io/milvus] Flat Index Use GPU search (Discussion #37042)





 
https://milvus.io/docs/performance_faq.md#How-to-set-nlist-and-nprobe-for-IVF-indexes
 As a rule of thumb, the recommended value of nlist is 4 × sqrt(n), where n is the total number of entities in a segment.
 
The collection has only tens of thousands items, I think there is only one segment in the collection. Let's say the collection has 50000 items, the proper value of nlist is 4 * sqrt(50000) = 894. Usually we set the value as power of 2. So, we can set it to be 1024.
 
If the collection has millions of items, then we need to estimate row count per segment by vector's dimension.

**Reply by yhmo** [reply]
Maybe your collection is "Strong" consistency level. "Strong" level will add extra 200ms ~ 400ms to search latency.
![Screenshot from 2024-10-23 17-40-39](https://github.com/user-attachments/assets/bfbe0160-e0e9-417f-b95b-cb3d241cd25d)


The Attu by default uses "Bounded" consistency level to search.
![Screenshot from 2024-10-23 17-41-35](https://github.com/user-attachments/assets/6f3af653-4df5-4f36-8519-e6065bdc262a)

In python, you can specify the level by:
```
res = collection.search(
    data=vectors,
    anns_field="my_vector",
    param={},
    consistency_level = "Bounded",
    limit=10,
)
```

More info about consistency level: https://milvus.io/docs/consistency.md#Consistency-levels

**Comment by 1556900941lizerui** [comment]
Thanks for the reply. I adopted your suggestion and used the Bounded level for a slight improvement in speed. I also used GPU_INV_FLAT for searching. But it still fluctuates in the range of 100ms to 400ms.



I am using the API provided by Langchain's Milvus class method to search, but I found that its underlying layer still uses the pymilvus library to search. Does this have an impact?



------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "milvus-io/milvus"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年10月23日(星期三) 下午5:44
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [milvus-io/milvus] Flat Index Use GPU search (Discussion #37042)





 
Maybe your collection is "Strong" consistency level. "Strong" level will add extra 200ms ~ 400ms to search latency.
 Screenshot.from.2024-10-23.17-40-39.png (view on web)
 
The Attu by default uses "Bounded" consistency level to search.
 Screenshot.from.2024-10-23.17-41-35.png (view on web)
 
In python, you can specify the level by:
 res = collection.search(     data=vectors,     anns_field="my_vector",     param={},     consistency_level = "Bounded",     limit=10, )  
More info about consistency level: https://milvus.io/docs/consistency.md#Consistency-levels

**Reply by yhmo** [reply]
Langchain's Milvus class is a wrapper of pymilvus, no much difference in performance.
Attu only takes 10ms to search, which is expected performance for 10000 entities collection. I think something is wrong with your python script.

**Comment by 1556900941lizerui** [comment]
Thanks for your help. I finally confirmed that it was a data consistency problem. Langchain did not explicitly pass it through when searching. I solved the problem by modifying the source code. I have a new question, whether Milvus supports the flask+gevent framework. After using this combination, I found that the program's service is non-blocking, and I haven't found the reason yet.Below is the code I'm calling





from gevent.pywsgi import WSGIServer
from gevent import monkey
monkey.patch_all()
import grpc._cython.cygrpc
grpc._cython.cygrpc.init_grpc_gevent()


from milvus_function import MilvusKBService
from flask import Flask,request,jsonify
from langchain.schema import Document


app = Flask(__name__)
http_server = WSGIServer((args.host, args.port), app)
http_server.serve_forever()



------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "milvus-io/milvus"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年10月23日(星期三) 晚上6:27
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [milvus-io/milvus] Flat Index Use GPU search (Discussion #37042)





 
Langchain's Milvus class is a wrapper of pymilvus, no much difference in performance.
 Attu only takes 10ms to search, which is expected performance for 10000 entities collection. I think something is wrong with your python script.

**Reply by yhmo** [reply]
Flask + pymilvus is ok, read this discussion: https://github.com/milvus-io/milvus/discussions/18194

**Comment by 1556900941lizerui** [comment]
https://milvus.io/blog/Accelerating-Similarity-Search-on-Really-Big-Data-with-Vector-Indexing.md
I'm sorry. I saw in this document related test information about using GPU to test FLAT index. Does this mean that when the index is FLAT, the GPU can be used during vector search?




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "milvus-io/milvus"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年10月22日(星期二) 下午2:46
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [milvus-io/milvus] Flat Index Use GPU search (Discussion #37042)





 
Supported GPU indexes are listed here: https://milvus.io/docs/gpu_index.md#GPU-Index
 
No GPU_FLAT.
 There is a search parameter "nprobe" for GPU_IVF_FLAT, you can set "nprobe" to be equal to "nlist" to achieve a full scan like FLAT.
 
collection.create_index(field_name = "my_vector", index_params = {"index_type": "GPU_IVF_FLAT", "metric_type": "COSINE", "params": {"nlist": 64}})
 search_params = {         "metric_type": "COSINE",         "params": {"nprobe": 64}     } res = collection.search(     data=vectors,     anns_field="vector",     limit=limit,     param=search_params, )

**Reply by yhmo** [reply]
This article was written on December 05, 2019, it is about Milvus v1.x. In Milvus v1.x, both CPU index and GPU index share the same index type enum. In Milvus v2.x, the index type enum for GPU and CPU is different. No GPU_FLAT in v2.x.

**Comment by 1556900941lizerui** [comment]
OK，thanks.



------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2024年10月25日(星期五) 上午10:13
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [milvus-io/milvus] Flat Index Use GPU search (Discussion #37042)














 
This article was written on December 05, 2019, it is about Milvus v1.x. In Milvus v1.x, both CPU index and GPU index share the same index type enum. In Milvus v2.x, the index type enum for GPU and CPU is different. No GPU_FLAT in v2.x.

**Comment by 1556900941lizerui** [comment]
Since my service needs to be migrated to a server with relatively poor CPU performance, I am afraid that the performance of using FLAT brute force search will decrease. I want to use GPU_FLAT instead to achieve the same effect. I tried to understand the principle of INV_FLAT. When nlist When used the same as nprobe, the same results as brute force search can be achieved. So in this case, the results of setting nlist and nprobe to 1 and setting both nlist and nprobe to 4 * sqrt(n) are theoretically the same, right? My understanding is that when set to 1, there is no need to traverse more clusters and faster search performance can be achieved. Not sure if my understanding is correct, please give me an answer.



------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "milvus-io/milvus"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年10月25日(星期五) 上午10:13
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [milvus-io/milvus] Flat Index Use GPU search (Discussion #37042)





 
This article was written on December 05, 2019, it is about Milvus v1.x. In Milvus v1.x, both CPU index and GPU index share the same index type enum. In Milvus v2.x, the index type enum for GPU and CPU is different. No GPU_FLAT in v2.x.

**Reply by liliu-z** [reply]
Yes that is the truth. But why don't use `GPU_BRUTE_FORCE` directly?

**Reply by 1556900941lizerui** [reply]
Thanks for the reply. I haven't had time to understand the principle of this index before. By consulting the official documents, I found that it is like the GPU version of FLAT. It is also equivalent to the situation where GPU_INV_FLAT nlist and nprobe are both 1, but the performance will be better. There should be no problem with my understanding?

**Reply by liliu-z** [reply]
> Thanks for the reply. I haven't had time to understand the principle of this index before. By consulting the official documents, I found that it is like the GPU version of FLAT. It is also equivalent to the situation where GPU_INV_FLAT nlist and nprobe are both 1, but the performance will be better. There should be no problem with my understanding?

Yes, it is hard to say how better, but maybe slightly due to the different data arrangement.

**Reply by 1556900941lizerui** [reply]
OK, I've applied the method, thank you for your help
